### ISSUE TITLE
primary worker process failed to initialize (110: Connection timed out

### ISSUE THREAD

**Issue by kuachen** [F-Stack/f-stack#232]
start nginx error with lcore_mask=f ,worker_processes=4
2018/06/11 23:49:03 [error] 17360#0: primary worker process failed to initialize (110: Connection timed out)
```
system:centos 7 64bit
./cpu_layout.py 
======================================================================
Core and Socket Information (as reported by '/sys/devices/system/cpu')
======================================================================

cores =  [0, 1, 2, 3]
sockets =  [0]

       Socket 0  
       --------  
Core 0 [0]       
Core 1 [1]       
Core 2 [2]       
Core 3 [3]       
```

**Comment by whl739** [comment]
Configure `daemon off` in nginx.conf or just run `example/helloworld` to see if there are any errors.

**Comment by hit1943** [comment]
iI have encountered this problem before，and I found that the nic  was not recognized. So check your nic drivers. if you use a nic that does not belong to intel, you should check this: http://dpdk.readthedocs.io/en/v17.11/nics/index.html

**Comment by kuachen** [comment]
i run it in virtualbox with hostonly mode,
it's no problem start nginx with  lcore_mask=1 ,worker_processes=1,
but,now i configured daemon off and run with  lcore_mask=3 ,worker_processes=2，
EAL: Error - exiting with code: 1
  Cause: num_procs[2] bigger than max_rx_queues[1]

detail log：
[dpdk]: lcore_mask=3
[dpdk]: channel=4
[dpdk]: promiscuous=1
[dpdk]: numa_on=1
[dpdk]: tso=0
[dpdk]: vlan_strip=1
[dpdk]: port_list=0
[port0]: addr=192.168.1.107
[port0]: netmask=255.255.255.0
[port0]: broadcast=192.168.1.255
[port0]: gateway=192.168.1.1
[freebsd.boot]: hz=100
[freebsd.boot]: fd_reserve=1024
[freebsd.boot]: kern.ipc.maxsockets=262144
[freebsd.boot]: net.inet.tcp.syncache.hashsize=4096
[freebsd.boot]: net.inet.tcp.syncache.bucketlimit=100
[freebsd.boot]: net.inet.tcp.tcbhashsize=65536
[freebsd.boot]: kern.ncallout=262144
[freebsd.sysctl]: kern.ipc.somaxconn=32768
[freebsd.sysctl]: kern.ipc.maxsockbuf=16777216
[freebsd.sysctl]: net.link.ether.inet.maxhold=5
[freebsd.sysctl]: net.inet.tcp.fast_finwait2_recycle=1
[freebsd.sysctl]: net.inet.tcp.sendspace=16384
[freebsd.sysctl]: net.inet.tcp.recvspace=8192
[freebsd.sysctl]: net.inet.tcp.nolocaltimewait=1
[freebsd.sysctl]: net.inet.tcp.cc.algorithm=cubic
[freebsd.sysctl]: net.inet.tcp.sendbuf_max=16777216
[freebsd.sysctl]: net.inet.tcp.recvbuf_max=16777216
[freebsd.sysctl]: net.inet.tcp.sendbuf_auto=1
[freebsd.sysctl]: net.inet.tcp.recvbuf_auto=1
[freebsd.sysctl]: net.inet.tcp.sendbuf_inc=16384
[freebsd.sysctl]: net.inet.tcp.recvbuf_inc=524288
[freebsd.sysctl]: net.inet.tcp.sack.enable=1
[freebsd.sysctl]: net.inet.tcp.blackhole=1
[freebsd.sysctl]: net.inet.tcp.msl=2000
[freebsd.sysctl]: net.inet.tcp.delayed_ack=0
[freebsd.sysctl]: net.inet.udp.blackhole=1
[freebsd.sysctl]: net.inet.ip.redirect=0
EAL: Probing VFIO support...
EAL:   no supported IOMMU extensions found!
EAL: VFIO support could not be initialized
EAL: PCI device 0000:00:03.0 on NUMA socket -1
EAL:   Invalid NUMA socket, default to 0
EAL:   probe driver: 8086:100e net_e1000_em
EAL: PCI device 0000:00:08.0 on NUMA socket -1
EAL:   Invalid NUMA socket, default to 0
EAL:   probe driver: 8086:100e net_e1000_em
EAL: PCI device 0000:00:09.0 on NUMA socket -1
EAL:   Invalid NUMA socket, default to 0
EAL:   probe driver: 8086:100e net_e1000_em
EAL: PCI device 0000:00:0a.0 on NUMA socket -1
EAL:   Invalid NUMA socket, default to 0
EAL:   probe driver: 8086:100e net_e1000_em
lcore: 0, port: 0, queue: 0
create mbuf pool on socket 0
create ring:dispatch_ring_p0_q0 success, 2047 ring entries are now free!
create ring:dispatch_ring_p0_q1 success, 2047 ring entries are now free!
EAL: Error - exiting with code: 1
  Cause: num_procs[2] bigger than max_rx_queues[1]

**Comment by whl739** [comment]
@kuachen 
Obviously, your nic has only one rx queue, so you couldn't start more than one process.

**Comment by kuachen** [comment]
@whl739 
so  if the nic only has  a few of rx queues，we  couldn't use f-stack improve performance  ？

**Comment by ygm521** [comment]
@kuachen did you solved it ?? i run on vmware centos7.2 the same err.
https://github.com/F-Stack/f-stack/issues/279

**Comment by mulhome-com** [comment]
@kuachen I run on ubuntu 20.04 and get the same following err. 
Can you tell me how to fix it ?

EAL: Detected CPU lcores: 40
EAL: Detected NUMA nodes: 2
EAL: Detected static linkage of DPDK
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'PA'
EAL: No available 1048576 kB hugepages reported
EAL: VFIO support initialized
EAL: Probe PCI driver: net_ice (8086:1593) device: 0000:05:00.0 (socket 0)
ice_load_pkg_type(): Active package is: 1.3.30.0, ICE COMMS Package (single VLAN mode)
TELEMETRY: No legacy callbacks, legacy socket not created
2023/05/10 12:13:18 [alert] 922017#0: fstack_conf null
2023/05/10 12:13:18 [error] 922016#0: primary worker process failed to initialize (110: Connection timed out)

**Comment by jfb8856606** [comment]
The error `primary worker process failed to initialize (110: Connection timed out)` when using `lcore_mask=f` (4 cores) with `worker_processes=4` is caused by a NIC hardware limitation — the NIC does not have enough RX queues to support the number of worker processes requested.

As seen in the related discussion (e.g., #234), F-Stack requires **one RX queue per worker process (lcore)**. If your NIC only supports 1 RX queue, you can only run a single worker process regardless of `lcore_mask`.

**How to check your NIC queue count:**
```bash
ethtool -l <interface_name>
```

**Solution:**
- Use a NIC that supports multiple RX/TX queues (at least as many as your intended worker count).
- In VM environments (VirtualBox, VMware), ensure the virtual NIC is configured with multi-queue support.
- Match `worker_processes` and the number of bits set in `lcore_mask` to the actual available RX queues.

Closing as **environment/configuration issue**. Please open a new issue if you encounter this on a multi-queue NIC with details of your hardware and configuration.
